### PR TITLE
[ShaderGraph] [bugfix 1352662] [2022.1] Use VPOS for fragment stage screenspace calculations

### PR DESCRIPTION
**Pull request by cdxntchou**
### Purpose of this PR
Fix for https://fogbugz.unity3d.com/f/cases/1352662/

2022.1:
 + https://github.com/Unity-Technologies/Graphics/pull/5428
 + https://github.com/Unity-Technologies/Graphics/pull/5784

2021.2 backport:
 + https://github.com/Unity-Technologies/Graphics/pull/5817

The issue is that the calculation for screenspace position is calculated from world space position via applying the matrix transforms.  However, the calculation is prone to floating point error, that is exacerbated greatly when the camera is far from the origin and the pipeline is not using camera-relative "world" space.

You can see in the examples below that the error can easily grow to be larger than a pixel.

This PR fixes this by changing the calculation in the fragment shader, to be based on the value returned by the VPOS fragment input.  This value is not based on the projection matrices, independent of camera position, and much more accurate overall.

Docs ticket: https://jira.unity3d.com/browse/GSG-593
Console Test PR: https://github.cds.internal.unity3d.com/unity/ScriptableRenderPipelinePrivate/pull/239

Test example:
Left sphere visualizes frac(screenspace pixel coordinate) using a gradient (red, black, green) where 0.5 is black.
Right sphere shows a screenspace checkerboard alphatest pattern

Before:  (at 20000 distance from origin) -- quite a bit of inaccuracy
image

After:   (at 20000 distance from origin) -- perfect 0.5 centered pixel coordinates everywhere!
image

Implementation Details:
We introduce two new global values that ShaderGraph can produce, and nodes can request, in addition to ScreenPosition.
The values are defined as:
PixelPosition                 2D pixel coordinate of the current fragment or vertex. Pixels are centered at 0.5.
NDCPosition                2D coordinate (0,1) across the viewport; equal to PixelPosition / ScreenResolution.
ScreenPosition (raw)     4D Clip Space homogenous coordinates of the current fragment or vertex.

Both PixelPosition and NDCPosition are calculated from VPOS in the fragment stage, and are calculated from ScreenPosition in other stages.

---
### Testing status

Tested in both HDRP and URP:
* Tested that Screen Position node and slots work correctly in both fragment and vertex shader.
* Validated that "Raw" mode works identically to previous, returning clip space ZW.
* Tested that the other modes return exact values in the fragment shader, based off of the VPOS value.
* Tested that VFX screen position works.

Added a Yamato test to validate screen position accuracy in fragment shader.  Previously you would have a good amount of error in the screen position from the floating point roundoff on the matrix transform calculations.  Now it gives exact pixel coordinate values out to at least 3 fractional digits.

Updated reference images.  Note that the new accuracy actually changes the test results for the dither node, because the position is slightly different (now more accurate), and the specific test we use is extremely position sensitive.

Master base branch:  35e38b19cc9212a9026b7f12d92e9fa79e677cfe

Yamato:
ShaderGraph: 🟢 
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252F1352662/.yamato%252Fall-shadergraph.yml%2523ShaderGraph_2021.2/8408772/job/pipeline
master:
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fall-shadergraph.yml%2523ShaderGraph_2021.2/8364150/job/pipeline

HDRP: 🟢
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252F1352662/.yamato%252Fall-hdrp.yml%2523PR_HDRP_trunk/8681927/job/pipeline

URP: 🟡 -- failures match master
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252F1352662/.yamato%252Fall-urp.yml%2523PR_URP_trunk/8702585/job/pipeline
master:
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fall-urp.yml%2523PR_URP_trunk/8678639/job/pipeline

Failing jobs:
Build URP_Terrain on iPhone_Metal_il2cpp_Linear_Standalone_build_Player on version trunk -- both failed non-test-related 🟡 
Build URP_Terrain on Android_Vulkan_il2cpp_Linear_Standalone_build_Player on version trunk -- both failed non-test-related 🟡
Build URP_Terrain on Android_OpenGLES3_il2cpp_Linear_Standalone_build_Player on version trunk -- both failed non-test-related 🟡
Build URP_Terrain on Linux_Vulkan_mono_Linear_Standalone_build_Player on version trunk -- both failed non-test-related 🟡
Build URP_Terrain on Win_Vulkan_mono_Linear_Standalone_build_Player on version trunk -- both failed non-test-related 🟡
Build URP_Terrain on Win_DX12_mono_Linear_Standalone_XR_build_Player on version trunk -- both failed non-test-related 🟡
Build URP_Terrain on Win_DX12_mono_Linear_Standalone_build_Player on version trunk -- both failed non-test-related 🟡
Build URP_Terrain on Win_DX11_mono_Linear_Standalone_XR_build_Player on version trunk -- both failed non-test-related 🟡
Build URP_Terrain on Win_DX11_mono_Linear_Standalone_build_Player on version trunk -- both failed non-test-related 🟡
URP_Foundation on iPhone_Metal_Standalone_il2cpp_Linear on version trunk -- both failed same 4 tests 🟡
URP_Lighting on iPhone_Metal_Standalone_il2cpp_Linear on version trunk -- both failed same 1 test 🟡
URP_Lighting on Android_Vulkan_Standalone_il2cpp_Linear on version trunk -- both failed on re-run (nondeterministic) 🟡
URP_Terrain on OSX_Metal_playmode_mono_Linear on version trunk -- both failed same 19 tests 🟡
URP_Terrain on Win_DX11_playmode_XR_mono_Linear on version trunk -- both failed same 13 tests 🟡
URP_PostPro on Win_DX11_playmode_XR_mono_Linear on version trunk -- both failed on re-run (nondeterministic) 🟡
ShaderGraph on OSX_Metal_playmode_mono_Linear on version trunk -- success on re-run 🟢

Console Tests: 🟡  -- failures match master

Console Test PR: https://github.cds.internal.unity3d.com/unity/ScriptableRenderPipelinePrivate/pull/239

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

**Comment by sebastienlagarde** [comment]
Hi, is this PR targetting 21.2 ? in this case it will need a backport

File: com.unity.shadergraph/Editor/Generation/Processors/GenerationUtils.cs [hunk 1]
```diff
@@ -746,6 +752,12 @@ internal static void GenerateSurfaceInputStruct(ShaderStringBuilder sb, ShaderGr
                 if (requirements.requiresScreenPosition)
                     sb.AppendLine("float4 {0};", ShaderGeneratorNames.ScreenPosition);
 
+                if (requirements.requiresNDCPosition)
+                    sb.AppendLine("float2 {0};", ShaderGeneratorNames.NDCPosition);
+
+                if (requirements.requiresPixelPosition)
+                    sb.AppendLine("float2 {0};", ShaderGeneratorNames.PixelPosition);
+
```
Comment (jessebarker) [inline com.unity.shadergraph/Editor/Generation/Processors/GenerationUtils.cs:760]: This looks fine, but why was ScreenPosition a float4?

**Review by jessebarker — APPROVED** [review]
LGTM

**Review by RemyUnity — APPROVED** [review]
Looking at the description and what has been tested, it seems ok to me.
Thanks for the detailed explanations, screenshots and the list of tests.

**Review by julienf-unity — APPROVED** [review]

**Review by ellioman — APPROVED** [review]

**Review by ernestasKupciunas — APPROVED** [review]
Approving without testing. Developer did a great job by testing these changes. New tests added. Good and clear description. Good job!

**Comment by PaulDemeulenaere** [comment]
This changes leads to a new compilation failure in VFX_URP test: 
```Unhandled log message: '[Error] Shader error in 'Hidden/VFX/SampleScene/System (1)/Output Particle Quad': 'VFXGetPositionRWS': no matching 1 parameter function at line 1647 (on d3d11)```
**Before merge**
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fvfx_urp-win-dx11.yml%2523VFX_URP_Win_DX11_playmode_mono_Linear_trunk/8938317/job
**After merge**
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fvfx_urp-win-dx11.yml%2523VFX_URP_Win_DX11_playmode_mono_Linear_trunk/8938121/job
I'm on it ⏳

**Comment by PaulDemeulenaere** [comment]
⚠️ While backporting this change, we must include 32c67d7276e6366447716179ab5e0eafe84b8fa3 ⚠️

**Comment by PaulDemeulenaere** [comment]
⚠️  This change 11417d30e2ebf958431796c3eee10595d3c5369f must be backported right after or preferably within the same backport PR. See https://github.com/Unity-Technologies/Graphics/pull/5784 ⚠️

**Pull request merged**